kube/src/api/dynamic.rs

```diff
@@ -188,6 +188,115 @@ impl TryFrom<DynamicResource> for Resource {
     }
 }
 
+/// Cheaply `Clone`-able string.
+// `String`: cloning requires additional allocation.
+// `StringRef::Dynamic`: cloning is one atomic operation.
+// `StringRef::Static`: cloning is essentially free.
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+enum StringRef {
+    /// Should be used when string is known during compilation.
+    Static(&'static str),
+    /// Should be used when string is only known at runtime
+    Dynamic(Arc<str>),
+}
+
+impl StringRef {
+    fn as_str(&self) -> &str {
+        match self {
+            StringRef::Static(s) => *s,
+            StringRef::Dynamic(s) => &*s,
+        }
+    }
+}
+
+/// Represents a type-erased object kind
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub struct GroupVersionKind {
+    /// API group
+    group: StringRef,
+    /// Version
+    version: StringRef,
+    /// Kind
+    kind: StringRef,
+    /// Concatenation of group and version
+    api_version: StringRef,
+}
+
+impl GroupVersionKind {
+    /// Creates `GroupVersionKind` from group, version and kind.
+    /// For `core` resources, group should be empty.
+    /// `api_version` will be created based on group and version
+    pub fn from_dynamic_gvk(group: &str, version: &str, kind: &str) -> Self {
```

Comment:
these parameter could be `T` where `<T: Into<String>>` so that you could pass them owned from the example if you wished

Reply:
We store `Arc<str>` internally, and its heap allocation is not compatible with `String`. So even if the user provides us with an owned String, it will still be reallocated. That's why I don't see benefits in being generic over `T: Into<String>`.

Reply:
Ah, ok  that makes sense.